### ISSUE TITLE
Is FlashOptim compatible with ZeRO1/ZeRO2/ZeRO3 using deepspeed ？

### ISSUE THREAD

**Issue by nideyongbao** [databricks/flashoptim#4]
Thanks for your great work! 

I consider how to train in transformers models with deepspeed, code likes this：

```
accelerator = Accelerator(...)
model = Qwen3_5ForConditionalGeneration.from_pretrained(...)
optimizer = FlashAdamW(model.parameters(), ...)
train_dataloader = build_dataloader(...)
model, optimizer, train_dataloader = accelerator.prepare(model, optimizer, train_dataloader)

for batch in train_dataloader:
    with accelerator.accumulate(model):
        outputs = model(**batch)
        loss = outputs.loss
        accelerator.backward(loss)
        optimizer.step()
        optimizer.zero_grad()
```

**Comment by josejg** [comment]
We haven't tested the code with the deepspeed library. At a high level it should work (e.g. it works with FSDP2), but you might have validate that stuff like distributed checkpointing is handled correctly

**Comment by kashif** [comment]
I tested FlashOptim with `trl`/`transformers` + `accelerate` + DeepSpeed on 2 GPUs via ZeRO-1, ZeRO-2, and ZeRO-3 via the PR https://github.com/databricks/flashoptim/pull/5; all trained successfully in the no-offload setting. In the smoke test, all three produced the same 1-step loss.

 The main caveat is CPU offload. FlashOptim’s quantized optimizer-state and fused-kernel benefits require CUDA-resident parameter shards at optimizer step time, so CPU-offloaded parameter shards are not a recommended/supported mode. In the PR, I added a clearer early error for that case instead of letting it fail later with a low-level quantization error.

  I have *not* fully validated DeepSpeed checkpoint/resume yet, only training smoke tests.

  Also, DeepSpeed prints its usual warning about using an untested optimizer with ZeRO, but the training path itself worked in these tests.
